Enable settings via the QMK settings interface

drpngx opened the pull request:
See README, and PRs in keybard and Vial

drpngx committed:
Enable custom settings in Vial (not tested).

drpngx committed:
Implement automouse and scroll bits, fix base qid, flip achordion (untested).

drpngx committed:
Adding some thoughts.

drpngx committed:
Clarify global max issue.

drpngx committed:
Cleanup comments.

drpngx committed:
Fix fall-through.

drpngx committed:
Soft-code the automouse timeout values.

drpngx committed:
Implement reset for some values.

drpngx committed:
Bounds check the mh_timer_index.

drpngx committed:
Use signed values to allow for -1 (still nothing tested while I am traveling).

drpngx committed:
Correctly use unsigned over the wire, increment introspecter, and copy mh_timer values one by one.

drpngx committed:
Document the signed issue.

drpngx committed:
Document reset issue.

drpngx committed:
Update the size of the saved_value so that we save and reload correctly.

If the values are wrong, the settings will be reset after a reboot.

drpngx committed:
Expand GT and LT predicates to make future updates easier.

drpngx committed:
Enable configuration of layer HSVs as user settings.

drpngx committed:
Mark rarely used function cold so they are moved out of the hot paths.

ilc commented:
Comments:

1. This is trying to do too much in one PR.  There's concepts in here you can break apart and make easier to digest..
1.1. Resetting of settings: NACK.   We just don't do this unless we have do.  People have enough pain getting this stuff setup.  It may get better once Keybard has the right APIs, and then be acceptable. but remember this may be someone's main pointingg device... and we want to give them a chance to actually complete the update with their pointer.
1.2.  You double implemented something the keybard extensions do, change layer colors.  We should look into extending vial.json if we want to make dynamic UI elements or just merge in what Via3 does, and fork off of Vial forever.  (I am far from against this.  the actual Vial infra is very stable.  It is the QMK stuff that shakes all over, and I think we can abstract Vial much more than is done today.)

Overall, I think this is well intentioned, but maybe we want to take a bigger look around before merging.

ilc reviewed (commented):
Comments are in the top of the PR.

ilc reviewed (changes requested):
Comments are in the top of the PR.

drpngx commented:
Thanks for the review @ilc! @SiriusStarr to summarize: this PR does the following:
1. Adds the QMK settings of `global_values`
- add hooks in the main qmk firmware in `quantum/qmk_settings.[ch]`
- create a file called `user_settings` in `svalboard`, and implement the protocols for existing fields in `global_values`
- added some documentation
2. Add the automouse timeouts to `global_values`, and bumping version number

Action items:
- rollback the `reset` function to just NOOP. That'll simplify some added clutter about exporting init values in `svalboard.[ch]`
- rollback the layer colors as it is somewhere else
- update docs, such as they are
- if I must, split the automouse timeout stuff into a separate PR
- drpngx: read up on via3

@SiriusStarr let me know if you want me to do the rollbacks first if you want to look at less code.

drpngx commented:
After chatting with @ilc on discord, it emerged that we could potentially use `vial.json`, which contains the keymaps, to store the additional QMK settings that I put in `keyboard.json` (in the UX). That way, there would be no `svalboard` specific code in keybard or the vial GUI. They would just paste them from `vial.json`. This puts us further down on the vial-specific path, which is another discussion.

Or ChatGPT might be right and `keyboard.json` is can be stored in the keyboard. Will certainly look into this.

ilc commented:
> After chatting with @ilc on discord, it emerged that we could potentially use `vial.json`, which contains the keymaps, to store the additional QMK settings that I put in `keyboard.json` (in the UX). That way, there would be no `svalboard` specific code in keybard or the vial GUI. They would just paste them from `vial.json`. This puts us further down on the vial-specific path, which is another discussion.

Just because it says vial, doesn't mean it won't be part of anything we do.  I don't see escaping vial.json for a while.

> Or ChatGPT might be right and `keyboard.json` is can be stored in the keyboard. Will certainly look into this.

Usually, ChatGPT is dumb as a brick when it comes to QMK,

ilc commented:
Closing due to inactivity.
